ophyd/redis_signal.py

```diff
@@ -0,0 +1,292 @@
+    _metadata_keys = None
+    _core_metadata_keys = ('connected', 'timestamp')
+
+    def __init__(
+        self, key, *, r, initial_value=None, serializer_deserializer=None, name=None, timestamp=None, **kwargs
+    ):
+        if name is None:
+            name = key
+        super().__init__(name=name, **kwargs)
+        if not isinstance(r, redis.StrictRedis):
+            r = redis.StrictRedis(**r)
+        self._r = r
+        self._key = key
+        self._subscription_thread = None
+        self._pubsub = None
+
+        self._last_read = {"value": None, "timestamp": None}
+        self._default_sub = self.SUB_VALUE
+
+        if serializer_deserializer is None:
+            self._serializer = json.dumps
+            self._deserializer = json.loads
+        else:
+            self._serializer = serializer_deserializer[0]
+            self._deserializer = serializer_deserializer[1]
+
+        if initial_value is not None:
+            if not self._r.exists(self._key):
+                self.set(initial_value)
+
+        if timestamp is None:
+            timestamp = time.time()
+
+        try:
+            connected = self._r.ping()
+        except redis.ConnectionError:
+            connected = False
+
+        self._metadata = dict(
+            connected=connected,
+            # read_access=True,
+            # write_access=True,
+            timestamp=timestamp,
+            # status=None,
+            # severity=None,
+            # precision=None,
+        )
+
+    @property
+    def timestamp(self):
+        '''Timestamp of the readback value'''
+        return self._metadata['timestamp']
+
+    @property
+    def connected(self):
+        'Is the signal connected to its associated hardware, and ready to use?'
+        return self._metadata['connected']  # and not self._destroyed
+
+    def set(self, value):
+        """Set value of signal. Sets value of redis key to the serialized dictionary of value and timestamp.
+
+        Returns
+        -------
+        st : Status
+            The status object is set to finished on successful write to redis, or an exception is set if redis.ConnectionError is raised.
+        """
+        st = Status(self)
+        try:
+            server_time = self._r.time()
+            ts = server_time[0] + server_time[1] / 1000000
+            self._r.set(
+                self._key,
+                self._serializer({"value": value, "timestamp": ts}),
+            )
+        except redis.ConnectionError as e:
+            st.set_exception(e)
+        st.set_finished()
+        return st
+
+    def get(self):
+        return self.read()[self.name]['value']
+
+    def put(self, value):
+        self.set(value)
+
+    def read(self):
+        v = self._r.get(self._key)
+        if v is None:
+            raise NoKey
+
+        self._last_read = self._deserializer(v)
+
+        return {
+            self.name: self._last_read,
+        }
+
+    def describe(self):
+        val = self.read()
```

Comment:
Should this use `self._last_read` (assume it's been read once) or force a read every time?